messaging/broker.go

```diff
@@ -144,943 +222,944 @@ func (b *Broker) closeTopics() {
+	} else if err = os.MkdirAll(b.path, 0700); err != nil {
+		return fmt.Errorf("mkdir: %s", err)
 	}
-	return nil
-}
 
-// Join joins an existing cluster.
-func (b *Broker) Join(u *url.URL) error {
-	if err := b.log.Join(u); err != nil {
-		return fmt.Errorf("raft: %s", err)
+	// Read header frame.
+	var sz uint32
+	if err := binary.Read(r, binary.BigEndian, &sz); err != nil {
+		return fmt.Errorf("read header size: %s", err)
+	}
+	buf := make([]byte, sz)
+	if _, err := io.ReadFull(r, buf); err != nil {
+		return fmt.Errorf("read header: %s", err)
 	}
-	return nil
-}
 
-// Publish writes a message.
-// Returns the index of the message. Otherwise returns an error.
-func (b *Broker) Publish(m *Message) (uint64, error) {
-	buf, _ := m.MarshalBinary()
-	return b.log.Apply(buf)
+	// Decode header.
+	sh := &snapshotHeader{}
+	if err := json.Unmarshal(buf, &sh); err != nil {
+		return fmt.Errorf("decode header: %s", err)
+	}
+
+	// Close any topics which might be open and clear them out.
+	b.closeTopics()
+
+	// Copy topic files from snapshot to local disk.
+	for _, st := range sh.Topics {
+		t := NewTopic(st.ID, b.topicPath(st.ID))
+
+		// Create topic directory.
+		if err := os.MkdirAll(t.Path(), 0700); err != nil {
+			return fmt.Errorf("make topic dir: %s", err)
+		}
+
+		// Copy data from snapshot into segment files.
+		// We don't instantiate the segments because that will be done
+		// automatically when calling Open() on the topic.
+		for _, ss := range st.Segments {
+			if err := func() error {
+				// Create a new file with the starting index.
+				f, err := os.Create(t.segmentPath(ss.Index))
+				if err != nil {
+					return fmt.Errorf("open segment: %s", err)
+				}
+				defer func() { _ = f.Close() }()
+
+				// Copy from stream into file.
+				if _, err := io.CopyN(f, r, ss.Size); err != nil {
+					return fmt.Errorf("copy segment: %s", err)
+				}
+
+				return nil
+			}(); err != nil {
+				return err
+			}
+		}
+
+		// Open new empty topic file.
+		if err := t.Open(); err != nil {
+			return fmt.Errorf("open topic: %s", err)
+		}
+		b.topics[t.id] = t
+	}
+
+	// Set the highest seen index.
+	if err := b.setMaxIndex(sh.Index); err != nil {
+		return fmt.Errorf("set max index: %s", err)
+	}
+	b.index = sh.Index
+
+	return nil
 }
 
-// PublishSync writes a message and waits until the change is applied.
-func (b *Broker) PublishSync(m *Message) error {
-	// Publish message.
-	index, err := b.Publish(m)
+// reset removes all files in the broker directory besides the raft directory.
+func (b *Broker) reset() error {
+	// Open handle to directory.
+	f, err := os.Open(b.path)
 	if err != nil {
 		return err
 	}
+	defer func() { _ = f.Close() }()
 
-	// Wait for message to apply.
-	if err := b.Sync(index); err != nil {
+	// Read directory items.
+	fis, err := f.Readdir(0)
+	if err != nil {
 		return err
 	}
 
+	// Remove all files & directories besides raft.
+	for _, fi := range fis {
+		if fi.Name() == "raft" {
```

Comment:
I started moving it out of the `broker` directory but I didn't want to expose raft through the configuration. I'm open to moving it but this worked for now.